### PR TITLE
Arc 806 - add github client to Commits backfill.

### PR DESCRIPTION
**Pull request by joshkay10**
Use the GitHub client to request the commit data.

Existing request was 2 GraphQL calls, get default branch and get commits for default branch,
rewrote the GraphQL query to include defaultBranchRef instead of having a seperate request to fetch it.
Manual testing confirmed return of the same data.

Also takes away the assumption that no default branch response is master.

Did not convert to REST due to the task processor relying on the cursor from the graphQL request and felt that changing that was out of scope.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -9,19 +9,6 @@
 		"project": "tsconfig.json",
 		"sourceType": "module"
 	},
-	"settings": {
```
Comment (joshkay10) [inline .eslintrc.json:12]: settings was duplicated.

File: src/config/feature-flags.ts [hunk 2]
```diff
@@ -24,6 +24,7 @@ export enum BooleanFlags {
 	USE_NEW_GITHUB_CLIENT_FOR_BRANCHES = "use-new-github-client-for-branches",
 	USE_NEW_GITHUB_CLIENT_FOR_DISCOVERY = "use-new-github-client-for-discovery",
 	USE_NEW_GITHUB_CLIENT_FOR_DEPLOYMENTS = "use-new-github-client-for-deployments",
+	USE_NEW_GITHUB_CLIENT_FOR_BACKFILL = "use-new-github-client-for-backfill_z7ps8",
```
Comment (joshkay10) [inline src/config/feature-flags.ts:27]: id at the end is because i used Switcheroo to generate it.

File: src/github/client/github-queries.ts [hunk 3]
```diff
@@ -78,7 +78,50 @@ export const getPullRequests = `query ($owner: String!, $repo: String!, $per_pag
     }
   }`;
 
-export const getCommits = (includeChangedFiles?: boolean) => `query ($owner: String!, $repo: String!, $per_page: Int!, $cursor: String, $default_ref: String!) {
+export type getCommitsResponse = {
```
Comment (joshkay10) [inline src/github/client/github-queries.ts:81]: Didn't have an appropriate Octokit type that I could find, felt like a unique case so made a custom Type.

File: src/github/client/github-queries.ts [hunk 4]
```diff
@@ -78,7 +78,50 @@ export const getPullRequests = `query ($owner: String!, $repo: String!, $per_pag
...
+          edges
+        }
+      }
+    }
+  }
+};
+
+export const getCommitsQuery = (includeChangedFiles?: boolean) => `query ($owner: String!, $repo: String!, $per_page: Int!, $cursor: String) {
+    repository(owner: $owner, name: $repo){
+      defaultBranchRef {
```
Comment (joshkay10) [inline src/github/client/github-queries.ts:95]: other than `defaultBranchRef` replacing `ref(qualifiedName: $default_ref)` this query is exactly the same

File: src/sync/commits.ts [hunk 5]
```diff
@@ -3,10 +3,60 @@ import { GitHubAPI } from "probot";
...
+
+		return {
+			edges,
+			jiraPayload
+		};
+	}
+	return getCommitsOctoKit(logger, github, gitHubClient, jiraHost, repository, cursor, perPage) 
+};
+
+/*
```
Comment (joshkay10) [inline src/sync/commits.ts:54]: Everything below here is the original implementation, it was too messy to have conditions everywhere to use octokit vs client.... clean up should be simple though. Renamed a few things for clarity.

File: test/fixtures/api/graphql/commit-queries.ts [hunk 6]
```diff
@@ -28,12 +29,11 @@ const query = `query ($owner: String!, $repo: String!, $per_page: Int!, $cursor:
     }
   }`;
 
-export const commitsNoLastCursor = {
+
+export const commitsNoLastCursor = (variables) => ({
 	query,
-	variables: {
```
Comment (joshkay10) [inline test/fixtures/api/graphql/commit-queries.ts:33]: moved these values out of here and into the test file, was easier to understand test with this data available.....

File: test/unit/sync/commits.test.ts [hunk 7]
```diff
@@ -1,237 +1,245 @@
 /* eslint-disable @typescript-eslint/no-var-requires,@typescript-eslint/no-explicit-any */
-import { commitsNoLastCursor, commitsWithLastCursor, getDefaultBranch } from "../../fixtures/api/graphql/commit-queries";
```
Comment (joshkay10) [inline test/unit/sync/commits.test.ts:2]: Please take extra care looking through these tests. They were previously all skipped, have rebuilt them all based off the original test cases.

File: src/sync/commits.ts [hunk 8]
```diff
@@ -3,10 +3,60 @@ import { GitHubAPI } from "probot";
...
+// According to the logs, GraphQL queries sometimes fail because the "changedFiles" field is not available.
+// In this case we just try again, but without asking for the changedFiles field.
+const retryFetchCommits = async (logger: LoggerWithTarget, err, gitHubClient: GitHubClient, repoOwner: string, repoName: string, cursor?: string | number, perPage?: number): Promise<getCommitsResponse> => {
+	const changedFilesErrors = err.errors?.filter(e => e.message?.includes("The changedFiles count for this commit is unavailable"));
+
+	if (changedFilesErrors?.length) {
+		logger.info("retrying without changedFiles");
+		return await gitHubClient.getCommitsPage(false, repoOwner, repoName, perPage, cursor);
+	}
+	throw new Error(err);
```
Comment (joshkay10) [inline src/sync/commits.ts:18]: Not sure if this is correct, what I want to achieve is if the above error for changedFilesErrors doesn't exist then just to bubble up the error to be handled elsewhere

File: .eslintrc.json (hunk 1 above)
Comment (mboudreau) [reply .eslintrc.json:12]: huh, must have been a merge issue.

File: src/config/feature-flags.ts (hunk 2 above)
Comment (mboudreau) [reply src/config/feature-flags.ts:27]: ...that's kind of annoying

File: src/github/client/github-queries.ts (hunk 3 above)
Comment (mboudreau) [reply src/github/client/github-queries.ts:81]: Yeah, most graphql ones are going to be unique, sadly.

File: src/github/client/github-queries.ts [hunk 9]
```diff
@@ -78,7 +78,50 @@ export const getPullRequests = `query ($owner: String!, $repo: String!, $per_pag
     }
   }`;
 
-export const getCommits = (includeChangedFiles?: boolean) => `query ($owner: String!, $repo: String!, $per_page: Int!, $cursor: String, $default_ref: String!) {
+export type getCommitsResponse = {
+  repository: {
+    defaultBranchRef: {
+      target: {
+        history: {
+          edges
```
Comment (mboudreau) [inline src/github/client/github-queries.ts:86]: What's `edges`? It should have it's own type.

File: src/sync/commits.ts (hunk 8 above)
Comment (mboudreau) [reply src/sync/commits.ts:18]: Personally, I'd have this retry mechanism build into the `getCommitsPage` for everyone to use.

File: src/github/client/github-client.ts [hunk 10]
```diff
@@ -226,12 +224,22 @@ export default class GitHubClient {
...
-				owner: owner,
+				owner,
 				repo: repoName,
 				per_page: perPage,
 				cursor
 			});
 		return response?.data?.data;
 	}
 
+	public async getCommitsPage(includeChangedFiles: boolean, owner: string, repoName: string, perPage?: number, cursor?: string | number): Promise<getCommitsResponse> {
```
Comment (mboudreau) [inline src/github/client/github-client.ts:235]: I don't particularly agree with having `includeChangedFiles` at the start of the argument list.  The argument list is also starting to become unmanageable.  I suggest we use a single object type for all params.  `includeChangedFiles` should be options, defaulted to false.

File: src/github/client/github-queries.ts (hunk 9 above)
Comment (joshkay10) [reply src/github/client/github-queries.ts:86]: array of commit objects, will add a type against.

File: src/github/client/github-client.ts (hunk 10 above)
Comment (joshkay10) [reply src/github/client/github-client.ts:235]: ha, agreed, I had originally created a repoDetails object for exactly that but stripped it out again when I dropped the default ref value. But yes args are getting too long. I'll do some tidy up.

**Review by mboudreau — COMMENTED** [review]
A few things I think needs to be changed. Happy to pair on this if needed.

**Comment by joshkay10** [comment]
Pollinator staging green, Manual testing on stage working as expected.

**Review by mboudreau — APPROVED** [review]
Nice!

**Pull request merged**